swift/common/middleware/ephemeral_objects.py

```diff
@@ -0,0 +1,301 @@
+# Copyright (c) 2010-2020 OpenStack Foundation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This middleware intercepts requests about "epheremal" objects (especially
+empty objects created as MPU placeholders) and stores them in an alternative
+backend (Redis).
+
+The middleware should be placed near the end of the pipeline.
+"""
+
+import json
+import time
+
+from paste.deploy import loadwsgi
+from six.moves.urllib.parse import parse_qs
+from swift.common.oio_utils import RedisDb
+from swift.common.swob import Request, HTTPInternalServerError
+from swift.common.utils import get_logger, MD5_OF_EMPTY_STRING, \
+    parse_connection_string
+from swift.common.wsgi import loadcontext, PipelineWrapper
+
+MIDDLEWARE_NAME = 'ephemeral_objects'
+DEFAULT_LIMIT = 10000
+
+DELIMITER = '/'
+SWIFT_SOURCE = 'EO'
+
+# For each UploadID, two redis keys will be created:
+# - one entry in zset MPU_LIST:{ACCOUNT}:{CONTAINER/BUCKET}: it will be used
+# to check if UploadId is still present and to support list-multipart-uploads
+# - one key as MPU_HDRS:{ACCOUNT}:{CONTAINER/BUCKET}:{UPLOADID} to hold headers
+LST_PREFIX = 'EO_LIST:'
+HDR_PREFIX = 'EO_HDRS:'
+
+# HACK: When iterating in Redis with non recursive listing, we force the next
+# zrangebylex iteration to skip already known prefixes. This is done by
+# suffixing the last known element with the "unicode replacement character",
+# which is the last valid character.
+END_MARKER = u"\ufffd"
```

Comment:
From Unicode 6.0, the last code point is `\U0010ffff`.

Reply:
And the last **valid** character seems to be `\U0010fffd`.